packages/contracts/contracts/discovery/ProjectVoting.sol

```diff
@@ -0,0 +1,245 @@
+
+    /// index in the votes array to project address
+    mapping(uint256 => address) public votesIndexToProject;
+
+    /// project address to index  in the votes array
+    mapping(address => uint256) public projectToVotesIndex;
+
+    /// user => votes
+    mapping(address => uint256) public userVoteCount;
+
+    /// project address => votes
+    mapping(address => uint256) public projectVoteCount;
+
+    /// project address => votes
+    mapping(address => uint256) public weightedProjectVoteCount;
+
+    /// project => (user => voted)
+    mapping(address => mapping(address => bool)) public userVotesPerProject;
+
+    /// Project address to status
+    /// @dev only valid up to the last time there was a vote. Use getProject(address) instead.
+    mapping(address => ProjectStatus) projectStatuses;
+
+    function projectVoting_projects()
+        public
+        view
+        virtual
+        returns (address[] memory);
+
+    function projectVoting_votingPeriod()
+        public
+        view
+        virtual
+        returns (Period memory);
+
+    function projectVoting_singleSlotDuration()
+        public
+        view
+        virtual
+        returns (uint256);
+
+    function projectVoting_initialBonus() public view virtual returns (int256);
+
+    function projectVoting_finalBonus() public view virtual returns (int256);
+
+    enum ProjectStatus {
+        InProgress,
+        Won,
+        Lost
+    }
+
+    struct Vote {
+        uint256 originalIndex;
+        uint32 count;
+    }
+
+    function _vote(address projectAddress) internal {
+        require(
+            !userVotesPerProject[projectAddress][msg.sender],
+            "already voted in this project"
+        );
+        _defineWinners();
+        require(
+            projectStatuses[projectAddress] == ProjectStatus.InProgress,
+            "project is not in progress"
+        );
+        userVotesPerProject[projectAddress][msg.sender] = true;
+        userVoteCount[msg.sender]++;
+        projectVoteCount[projectAddress]++;
+        weightedProjectVoteCount[projectAddress] += _calculateWeightedVote(
+            block.timestamp
+        );
+
+        if (
+            votes.length == 0 ||
+            (projectToVotesIndex[projectAddress] == 0 &&
+                votesIndexToProject[0] != projectAddress)
+        ) {
+            votes.push(1);
+            votesIndexToProject[votes.length - 1] = projectAddress;
+            projectToVotesIndex[projectAddress] = votes.length - 1;
+        } else {
+            votes[projectToVotesIndex[projectAddress]]++;
+        }
+    }
+
+    function _defineWinners() internal {
+        address[] memory computedWinners = _getWinners();
+
+        if (computedWinners.length == winners.length) {
+            return;
+        }
+
+        for (uint256 i = 0; i < computedWinners.length; i++) {
+            winners.push(computedWinners[i]);
+            projectStatuses[computedWinners[i]] = ProjectStatus.Won;
+        }
```

Comment:
isn't this loop going to be inserting duplicate records into `winners`? if this function ends up being called multiple times, for example, during two separate votes